### ISSUE TITLE
[Phase 2] Streaming Ingestion

### ISSUE THREAD

**Issue by ennajari** [CardiovasculaireFederated/cardiologie-project#13]
Ingestion des donnees en streaming

**Comment by ennajari** [comment]
Fermé par le commit 050873d - Pipeline de streaming et preprocessing implémenté

**Comment by ennajari** [comment]
<img width="701" height="2815" alt="Image" src="https://github.com/user-attachments/assets/fca11082-cabc-45bf-bb41-8db8a89e3eac" />

**Comment by ennajari** [comment]
### Le Pipeline Complet : De la Donnée Brute au Modèle Entraîné

<img width="961" height="6448" alt="Image" src="https://github.com/user-attachments/assets/2453ce6f-e6f0-4845-9c48-76f44e13d324" />

**Comment by ennajari** [comment]
**DATA SOURCE**

> Fichier : heart_disease.csv (10,000 lignes)


Contenu : Données brutes avec 21 colonnes non traitées

**SIMULATEUR (Data Injector)**

> Fichier : _simulator/data_injector.py_ Rôle : Simuler un flux de données en temps réel Fonctionnement :


DataInjector(batch_size=10, interval_seconds=5)
├─ Découpe le CSV en batches de 10 lignes
├─ Envoie un batch toutes les 5 secondes
└─ Crée des fichiers : heart_in_0.csv, heart_in_1.csv, ...
Output : Fichiers CSV dans data/incoming/

**SPARK STREAMING**

> Fichier : _client/client.py (méthode load_csv_batch)_ Rôle : Lire les fichiers CSV en DataFrames Spark Fonctionnement :


```
_spark.read.csv(
    path="data/incoming/heart_in_0.csv",
    header=True,
    schema=get_heart_schema()  # Validation stricte des types
)_
```


**PREPROCESSING PIPELINE**

> Fichier : _client/preprocessing.py_ C'est LE CŒUR du système ! Pipeline Spark ML en 5 étapes :


**ÉTAPE 1 : Nettoyage (clean_missing_values)**

Colonnes numériques avec NaN → Moyenne de la colonne
Colonnes catégorielles avec NaN → "Unknown"
**ÉTAPE 2 : Encodage Target**

"Heart Disease Status":
  - "Yes" → 1.0
  - "No"  → 0.0
→ Nouvelle colonne: "label"
**ÉTAPE 3 : Encodage Catégoriel (StringIndexer)**

11 colonnes catégorielles transformées:
  Gender: "Male"/"Female" → 0/1
  Smoking: "Yes"/"No" → 0/1
  Exercise Habits: "High"/"Medium"/"Low" → 0/1/2
  ...
→ 11 nouvelles colonnes: Gender_indexed, Smoking_indexed, ...
**ÉTAPE 4 : Vectorisation (VectorAssembler)**

Entrée: 
  - 9 colonnes numériques (Age, BMI, ...)
  - 11 colonnes catégorielles encodées (*_indexed)

Assemblage:
  [num1, num2, ..., num9, cat1, cat2, ..., cat11]
  
Output:
  - Colonne "features_raw": DenseVector de 20 éléments
**ÉTAPE 5 : Normalisation (StandardScaler)**

Pour chaque feature i:
  x_norm[i] = (x[i] - mean[i]) / std[i]

Résultat:
  - mean = 0
  - std = 1
  - Colonne "features": DenseVector[20] normalisé

> Output Final : DataFrame avec 2 colonnes : features (Vector[20]) et label (Double)


**_CONVERSION SPARK → PYTORCH_**

> Fichier : client/client.py (méthode spark_to_pytorch) Rôle : Transformer les données Spark en tenseurs PyTorch Fonctionnement :


**# 1. Collecter les données depuis Spark**
`rows = df.collect()  # Spark → Python`

**# 2. Extraire les vecteurs**
```
for row in rows:
    features = row['features'].toArray()  # DenseVector → NumPy
    X_list.append(features)
```

**# 3. Convertir en PyTorch**
X_tensor = torch.from_numpy(np.array(X_list, dtype=float32))
y_tensor = torch.from_numpy(np.array(y_list, dtype=float32))
Output :
X: Tensor de shape (N, 20)
y: Tensor de shape (N,)

**MODÈLE PYTORCH**
Fichier : _client/model.py_ Architecture :

HeartDiseaseModel:
```
  Input: 20 features
  ├─ FC(20 → 64) + BatchNorm + ReLU + Dropout(0.2)
  ├─ FC(64 → 32) + BatchNorm + ReLU + Dropout(0.2)
  ├─ FC(32 → 16) + BatchNorm + ReLU + Dropout(0.1)
  └─ FC(16 → 1)  → Logit (avant sigmoid)
```


**ENTRAÎNEMENT**

> Fichier : _client/client.py_ (méthode train_on_batch) Fonctionnement :


```
for epoch in range(5):
    # Forward
    outputs = model(X_tensor)  # Shape: (N, 1)
    loss = criterion(outputs.squeeze(), y_tensor)
    
    # Backward
    optimizer.zero_grad()
    loss.backward()
    optimizer.step(
```)

> Résultat : Loss décroissante (0.7253 → 0.5744)


**SAUVEGARDE**
Fichier : client/client.py (méthode save_processed_batch) Format : Fichiers .pt PyTorch

```
torch.save({
    'X': X_tensor,  # Tensor(N, 20)
    'y': y_tensor   # Tensor(N)
}, 'data/processed/batch_0.pt')
```

**Comment by ennajari** [comment]
### Résumé : Pourquoi c'est un "Pipeline" ?
Un pipeline c'est une chaîne de transformations automatiques où :

1. Chaque étape transforme les données d'un format à un autre
2. L'ordre est important : on ne peut pas normaliser avant de vectoriser
3. Tout est automatisé : une fois configuré, le pipeline traite n'importe quel batch
4. Réutilisable : Le pipeline est "fité" une fois sur le dataset complet, puis réappliqué à chaque nouveau batch

Analogie :

- Comme une chaîne de production automobile :
- Matière première → Forge → Assemblage → Peinture → Contrôle qualité → Produit fini

Ici :

- CSV brut → Nettoyage → Encodage → Vectorisation → Normalisation → Tensor prêt pour l'IA
